lib/client/clientcache/clientcache.go

```diff
@@ -184,21 +212,54 @@ func (c *Cache) Clear() error {
 	return trace.NewAggregate(errors...)
 }
 
-func (c *Cache) addToCache(k key, clusterClient *client.ClusterClient) {
+func (c *Cache) clearForKey(k key) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	clt, ok := c.clients[k]
+	delete(c.clients, k)
+	if ok {
+		err := clt.client.Close()
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+
+func (c *Cache) addToCache(k key, cc *clientWithCert) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	c.clients[k] = clusterClient
+	c.clients[k] = cc
 }
 
-func (c *Cache) getFromCache(k key) *client.ClusterClient {
+func (c *Cache) getFromCache(k key) *clientWithCert {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
 	clt := c.clients[k]
 	return clt
 }
 
+func (c *Cache) isCoreTLSCertUnchanged(ctx context.Context, profile string, coreTLSCertFromCache []byte) (bool, error) {
+	tc, err := c.cfg.NewClientFunc(ctx, profile, "")
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	keyRing, err := tc.LocalAgent().GetCoreKeyRing()
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	if bytes.Equal(coreTLSCertFromCache, keyRing.TLSCert) {
```

Comment:
It seemed to me that comparing the TLS cert is the easiest way to detect if a new client should be created, but I'm open to suggestions.